xxxx-ledger-plugin-interface-2/xxxx-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,502 @@
+
+## Class: Transfer
+<code>class Transfer</code>
+
+The `Transfer` class is used to describe transfers from the originator of the sendTransfer call towards some `destination`. All fields are required and MUST NOT be undefined. However, `amount` MAY be the value `'0'`, `data` MAY be an empty buffer and `custom` MAY be an empty object.
+
+###### Fields
+| Type | Name | Description |
+|:--|:--|:--|
+| `String` | [amount](#transferamount) | Integer transfer amount, in the ledger's base unit |
+| `String` | [destination](#transferdestination) | ILP destination address |
+| `Buffer` | [data](#transferdata) | Transport-layer data |
+| `String` | [executionCondition](#transferexecutioncondition) | Cryptographic hold condition |
+| `String` | [expiresAt](#transferexpiresat) | Expiry time of the cryptographic hold |
+| `Object` | [custom](#transfercustom) | Object containing ledger plugin specific options |
+
+###### Example
+``` js
+{
+  amount: '100',
+  destination: 'example.us.acmebank.bob',
+  executionCondition: 'I3TZF5S3n0-07JWH0s8ArsxPmVP6s-0d0SqxR6C3Ifk',
+  expiresAt: '2017-12-02T11:51:26.627Z',
+  data: Buffer.alloc(0),
+  custom: {
+    _alternateAccount: 'bob-savings',
+    executionPriority: 9
+  }
+}
+```
+
+### Fields
+
+#### Transfer#amount
+<code>**amount**:String</code>
+
+An integer amount, represented as a string of base-ten digits. MUST be in the range `0..9223372036854775807` (`>= 0` and `< 2^64`).
+
+###### Example
+
+``` js
+'100'
+```
+
+#### Transfer#destination
+<code>**destination**:String</code>
+
+An [ILP address](../0003-interledger-protocol/), denoting the payment's final destination. Length MUST be in the range `1..1023` (`> 0` and `< 2^10`).
+
+If the `destination` address is too long, the ledger plugin MUST reject with a `MaximumIlpAddressLengthExceededError`.
+
+###### Example
+
+``` js
+'example.us.acmebank.bob'
+```
+
+#### Transfer#data
+<code>**data**:Buffer</code>
+
+Arbitrary transport-layer data that travels with the payment. Size MUST be in the range `0..32767` (`>= 0` and `< 2^15`).
+
+If the data is too large, the ledger plugin MUST reject with a `MaximumDataSizeExceededError`.
+
+#### Transfer#executionCondition
+<code>**executionCondition**:String</code>
+
+A cryptographic challenge used for implementing holds. The underlying ledger MUST hold the transfer until the condition has been fulfilled or the `expiresAt` time has been reached.
+
+Conditions are the base64url-encoded SHA-256 hash of a random or pseudo-random 32-byte preimage called the fulfillment.
+
+###### Example
+
+``` js
+'I3TZF5S3n0-07JWH0s8ArsxPmVP6s-0d0SqxR6C3Ifk'
+```
+
+#### Transfer#expiresAt
+<code>**expiresAt**:String</code>
+
+An ISO 8601 datetime string representing the expiry date for the transfer. Must include the UTC timezone identifier `Z`.
+
+This date MUST be in the future, otherwise the plugin MUST reject with a `UnacceptableExpiryError`.
+
+###### Example
+
+``` js
+'2017-12-02T11:51:26.627Z'
+```
+
+#### Transfer#custom
```

Comment:
It's not clear to me how this differs from `data` if the fields MUST be passed to the receiver.

My assumption is that the plugin will know what needs to go to the receiver and what doesn't based on the underlying connection type.

Reply:
> It's not clear to me how this differs from data

@emschwartz is correct.

`data` is passed on by connectors and plugins. `custom` is only passed on by plugins (generally speaking), but not by connectors. Although there are exceptions. (For example, I'm using `custom.ilpLegacyAmount` to pass on the ILP amount in my connector refactor.)

I'll see if I can improve the wording.

Reply:
I worry about having something that is intended for the next connector (local receiver) mixed in with something for the plugin. 

This implies the plugin must know how to differentiate between the two (and only send the data intended for the next connector) and that the caller must have some knowledge of the underlying plugin (i.e. does it support X in `custom`).

This also suggests that we've simply moved the implementation challenges of `sendMessage` into `sendTransfer`. If I have a ledger that doesn't support messaging I need to establish a communications channel specifically for sending this data in parallel to the transfers.

Could we split `custom` into `pluginData` and `peerData`?

As a caller, if I have no idea what plugin I am talking to, I can still safely put anything serializable into `peerData` and know it will get to my peer.

I will only use `pluginData` if I know what plugin I am speaking to or if some of the data in there becomes standard and I expect all plugins to understand it.